### PR TITLE
[NodeSearchBundle][SearchBundle] Multi domain/language search population fix

### PR DESCRIPTION
**Pull request by cv65kr**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets | #1246

Now we can use language names like "en", "en_US", "en_us", also fixes error mentioned in #1246 ticket.

--edited by numkil
Also with new feature to check for unavailable language analyzers and prompt the user.

### CODE REVIEW

File: src/Kunstmaan/NodeSearchBundle/Configuration/NodePagesConfiguration.php [hunk 1]
```diff
@@ -137,14 +136,23 @@ public function createIndex()
         );
 
         foreach ($this->locales as $locale) {
-            $localeAnalysis = clone($analysis);
-            $language = $this->analyzerLanguages[$locale]['analyzer'];
+            // Multilanguage check
+            if ( count(preg_grep('/[a-z]{2}_?+[a-zA-Z]{2}/', [$locale])) > 0 ) {
```
Comment (sandergo90) [inline src/Kunstmaan/NodeSearchBundle/Configuration/NodePagesConfiguration.php]: @cv65kr can you remove the space before "count" ?

File: src/Kunstmaan/NodeSearchBundle/Configuration/NodePagesConfiguration.php [hunk 2]
```diff
@@ -318,7 +326,7 @@ public function setAnalysis(Index $index, AnalysisFactoryInterface $analysis)
      *
      * @return Mapping
      */
-    protected function createDefaultSearchFieldsMapping(Index $index, $lang = 'en')
+    protected function createDefaultSearchFieldsMapping(Index $index)
```
Comment (Numkil) [inline src/Kunstmaan/NodeSearchBundle/Configuration/NodePagesConfiguration.php]: Can you undo this change? I know it looks stupid but this is necessary to avoid BC breaks on older projects that still require $lang in this method.

File: src/Kunstmaan/NodeSearchBundle/Configuration/NodePagesConfiguration.php [hunk 3]
```diff
@@ -334,7 +342,7 @@ protected function createDefaultSearchFieldsMapping(Index $index, $lang = 'en')
      * @param Index  $index
      * @param string $lang
      */
-    protected function setMapping(Index $index, $lang = 'en')
+    protected function setMapping(Index $index, $lang)
```
Comment (Numkil) [inline src/Kunstmaan/NodeSearchBundle/Configuration/NodePagesConfiguration.php]: Can you undo this change? I know it looks stupid but this is necessary to avoid BC breaks on older projects that still require $lang in this method.

File: src/Kunstmaan/SearchBundle/Search/Search.php [hunk 4]
```diff
@@ -77,6 +77,8 @@ public function getClient()
      */
     public function createDocument($uid, $document, $indexName = '', $indexType = '')
     {
+        $indexName = strtolower($indexName);
```
Comment (Numkil) [inline src/Kunstmaan/SearchBundle/Search/Search.php]: Why are you doing strtolower here?

**Comment by Numkil** [comment]
hey @cv65kr you seem to have made a mistake in your rebase/squash procedure and your commit now contains commits from other people  not in this branch yet.

**Comment by cv65kr** [comment]
@Numkil I saw that, but i had to leave ;) i will try to fix that today

**Comment by Numkil** [comment]
No hurries I will review it when you are finished.

**Comment by cv65kr** [comment]
@Numkil done.
Going back to your question about `strtolower`, letters must be lowercase because elastic return exception if index name have a uppercase letters. Check out on my screenshot :)

elastic

File: src/Kunstmaan/NodeSearchBundle/Configuration/NodePagesConfiguration.php [hunk 5]
```diff
@@ -137,14 +136,23 @@ public function createIndex()
         );
 
         foreach ($this->locales as $locale) {
-            $localeAnalysis = clone($analysis);
-            $language = $this->analyzerLanguages[$locale]['analyzer'];
+            // Multilanguage check
+            if (count(preg_grep('/[a-z]{2}_?+[a-zA-Z]{2}/', [$locale])) > 0) {
```
Comment (Numkil) [inline src/Kunstmaan/NodeSearchBundle/Configuration/NodePagesConfiguration.php]: Why not use preg_match instead?

File: src/Kunstmaan/NodeSearchBundle/Configuration/NodePagesConfiguration.php [hunk 6]
```diff
@@ -137,14 +136,23 @@ public function createIndex()
...
-            //create index with analysis
-            $this->setAnalysis($index, $localeAnalysis->setupLanguage($language));
+            if (array_key_exists($locale, $this->analyzerLanguages)) {
+                $localeAnalysis = clone $analysis;
+                $language = $this->analyzerLanguages[$locale]['analyzer'];
+
+                // Create index with analysis
+                $this->setAnalysis($index, $localeAnalysis->setupLanguage($language));
+            } else {
+                $index->create([]);
```
Comment (Numkil) [inline src/Kunstmaan/NodeSearchBundle/Configuration/NodePagesConfiguration.php]: I would prefer a very clear warning happening here that says something like,
"hey we did not crash but something is wrong here, we could not set up any analysis for this language".

File: src/Kunstmaan/NodeSearchBundle/Configuration/NodePagesConfiguration.php (hunk 5 above)
Comment (cv65kr) [reply src/Kunstmaan/NodeSearchBundle/Configuration/NodePagesConfiguration.php]: good point

File: src/Kunstmaan/NodeSearchBundle/Configuration/NodePagesConfiguration.php (hunk 6 above)
Comment (cv65kr) [reply src/Kunstmaan/NodeSearchBundle/Configuration/NodePagesConfiguration.php]: @Numkil, You mean exception?

Comment (Numkil) [inline src/Kunstmaan/NodeSearchBundle/Configuration/NodePagesConfiguration.php]: I think the best course of action is to somehow show a warning in the command output. Otherwise it will not be clear for developers why their search might not be behaving as expected without manually checking the index metadata. If a simple message is shown in the console command output then at least users will immediately know that they forgot to define an analyzer language

**Comment by cv65kr** [comment]
@Numkil something like that?

**Comment by Numkil** [comment]
@cv65kr yes this is awesome! I saw one little typo where you typed answear instead of answer but if you fix that I will merge this PR tonight or tomorrow. Thank you for your help and quick responses!

**Comment by cv65kr** [comment]
@Numkil ohhh, already fixed :)

**Comment by Numkil** [comment]
@cv65kr thanks for the contribution.

**Comment by cv65kr** [comment]
@Numkil, no problem :) You can close #1246 because its solved by this PR.
